File: Programmers/[1844] 게임 맵 최단거리.java
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;

class Point {
    int y;
    int x;

    public Point(int y, int x) {
        this.y = y;
        this.x = x;
    }
}

class Solution {

    static final int[] dy = {-1, 0, 1, 0};
    static final int[] dx = {0, -1, 0, 1};
    static int[][] visited;
    static int rowLength;
    static int colLength;

    void bfs(int[][] maps) {
        Queue<Point> queue = new LinkedList<>();
        queue.offer(new Point(0, 0));
        visited[0][0] = 1;

        while (!queue.isEmpty()) {
            Point cur = queue.poll();
            for (int i = 0; i < 4; i++) {
                int ny = cur.y + dy[i];
                int nx = cur.x + dx[i];
                if (0 <= ny && ny < rowLength && 0 <= nx && nx < colLength) {
                    if (maps[ny][nx] == 1 && visited[ny][nx] == -1) {
                        visited[ny][nx] = visited[cur.y][cur.x] + 1;
                        queue.offer(new Point(ny, nx));
                        if (ny == rowLength - 1 && nx == colLength - 1) {
                            queue.clear();
                            break;
                        }
                    }
                }
            }
        }
    }

    public int solution(int[][] maps) {

        rowLength = maps.length;
        colLength = maps[0].length;
        visited = new int[rowLength][colLength];
        for (int[] row : visited) {
            Arrays.fill(row, -1);
        }

        bfs(maps);

        return visited[rowLength - 1][colLength - 1];
    }
}

File: Programmers/[1844] 게임 맵 최단거리.py
from collections import deque


def solution(maps):
    n = len(maps)
    m = len(maps[0])

    dy = [-1, 0, 1, 0]
    dx = [0, 1, 0, -1]

    isVisited = [[-1] * m for _ in range(n)]
    isVisited[0][0] = 0
    queue = deque([[0, 0]])
    while queue:
        y, x = queue.popleft()

        if y == n - 1 and x == m - 1:
            return isVisited[y][x] + 1

        for i in range(4):
            ny = y + dy[i]
            nx = x + dx[i]
            if 0 <= ny < n and 0 <= nx < m:
                if maps[ny][nx] == 1 and isVisited[ny][nx] == -1:
                    isVisited[ny][nx] = isVisited[y][x] + 1
                    queue.append([ny, nx])
    return -1
